File: elixir/lib/bottles.ex
defmodule Bottles do
  def song() do
    verses(99, 0)
  end

  def verses(start, stop) do
    start..stop
    |> Enum.map(&verse/1)
    |> Enum.join("\n")
  end

  def verse(n) do
    first_verse(n) <> second_verse(n)
  end

  defp first_verse(n) do
    "#{capitalize(remaining_bottles(n))} #{grammatic_number(n)} of milk on the wall, #{remaining_bottles(n)} #{grammatic_number(n)} of milk.\n"
  end

  defp second_verse(0) do
    "Go to the store and buy some more, 99 bottles of milk on the wall.\n"
  end

  defp second_verse(n) do
    "Take #{to_take_down(n)} down and pass it around, #{remaining_bottles(n - 1)} #{grammatic_number(n - 1)} of milk on the wall.\n"
  end

  defp grammatic_number(n) do
    if n == 1 do
      "bottle"
    else
      "bottles"
    end
  end

  defp capitalize(n) do
    if is_integer(n) do
      n
    else
      String.capitalize(n)
    end
  end

  defp remaining_bottles(n) do
    if n == 0 do
      "no more"
    else
      n
    end
  end

  def to_take_down(n) do
    if n > 1 do
      "one"
    else
      "it"
    end
  end
end

File: ruby/lib/bottles.rb
class Bottles
  def song
    verses(99, 0)
  end

  def verses(last_verse, first_verse)
    (first_verse..last_verse).map { |n| verse(n) }.reverse.join("\n")
  end
  def verse(n)
    first_line(n) + second_line(n)
  end

  private

  def first_line(n)
    if n > 0
      "#{n} #{grammatical_number(n)} of milk on the wall, #{n} #{grammatical_number(n)} of milk.\n"
    else
      "No more bottles of milk on the wall, no more bottles of milk.\n"
    end
  end

  def second_line(n)
    if n > 0
      "Take #{to_take_down(n)} down and pass it around, #{remaining_bottles(n)} of milk on the wall.\n"
    else
      "Go to the store and buy some more, 99 bottles of milk on the wall.\n"
    end
  end

  def grammatical_number(n)
    if n > 1
      "bottles"
    else
      "bottle"
    end
  end

  def to_take_down(n)
    if n == 1
      "it"
    else
      "one"
    end
  end

  def remaining_bottles(n)
    if n == 1
      "no more bottles"
    else
     "#{n-1} #{grammatical_number(n-1)}"
    end
  end
end
